### ISSUE TITLE
Empty voxels when bringing back softseg to native space

### ISSUE THREAD

**Issue by sandrinebedard** [sct-pipeline/contrast-agnostic-softseg-spinalcord#11]
## Description
For #2, we want to bring back the softseg to the native space due to reasons explained [here](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/pull/9#issuecomment-861824066). 

Right now, we use `sct_apply_transfo`with linear interpolation

https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/88fbfc399a8b1da9e08052c2a8437545d9a0bcbd/process_data.sh#L187

However, the resulting softseg sometime have empty voxels at the extremity of the segmentation while in T2w space, the segmentation is complete:

![image](https://user-images.githubusercontent.com/71230552/124797353-28b79700-df20-11eb-84f1-effd8ac2fec2.png)

I also openned it in fsleyes to make sure it wasn't just the threshold of 0.5 appplied in the qc report:
![image](https://user-images.githubusercontent.com/71230552/124794820-6b2ba480-df1d-11eb-9d6d-32a10659a49d.png)
(sub-amu01_acq-T1w_MTS)

The softseg in T2w space:
![image](https://user-images.githubusercontent.com/71230552/124797602-76cc9a80-df20-11eb-9ae5-6d195f784dd9.png)

This doesn't happen all the time, I am wondering if this won't be a problem when we have all the manual segmentations since registration is based on the segmentation and slices in the extremity aren't always good.

This is the segmentation with stc_deepseg_sc used here for sub-amu01_acq-T1w_MTS:
![image](https://user-images.githubusercontent.com/71230552/124797823-b7c4af00-df20-11eb-90ec-df2e129d5f88.png)

**Comment by jcohenadad** [comment]
Wow, that is a strange bug. I would start investigating what happens if you change `-x linear` with `-x nn` or `-x spline`.

Then, I would open the warping field and check if there are discontinuities at the location where you "loose" voxels. My gut feeling is that the warping field has some issues which we would need to address.

**Comment by sandrinebedard** [comment]
>  I would start investigating what happens if you change `-x linear` with `-x nn` or `-x spline`.

Same thing happens with `-x nn` and `-x spline`:

With `-x spline`: 
![image](https://user-images.githubusercontent.com/71230552/124940427-fa958e00-dfd7-11eb-858a-2fc92017075c.png)

With `-x nn`:

![image](https://user-images.githubusercontent.com/71230552/124940530-100ab800-dfd8-11eb-8bef-457fb2a04af5.png)

The problem must be with the warping field, I'll check for discontinuities!

**Comment by sandrinebedard** [comment]
Hmm, there are no discontinuities on the warping field where the empty voxels are...

![image](https://user-images.githubusercontent.com/71230552/124947684-100db680-dfde-11eb-87a3-98fc2bdfbf15.png)

![image](https://user-images.githubusercontent.com/71230552/124947731-19971e80-dfde-11eb-82c2-9177e699a2ad.png)


I also applied the transformation to the coverage image (image of 1 that we registered to T2w space) to check if the same thing happens and yes:

![image](https://user-images.githubusercontent.com/71230552/124947999-506d3480-dfde-11eb-93c5-3819258ab857.png)

**Comment by jcohenadad** [comment]
@sandrinebedard I'm having a hard time understanding what is the orientation on https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/11#issuecomment-876534386. Could you please indicate A/P/I/S/R/L in the figure, and also show a view of the spinal cord (maybe overlaid) so we can understand what location in the warping field corresponds to the discontinuity in the segmentation.

**Comment by sandrinebedard** [comment]
> @sandrinebedard I'm having a hard time understanding what is the orientation on [#11 (comment)](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/11#issuecomment-876534386). Could you please indicate A/P/I/S/R/L in the figure, and also show a view of the spinal cord (maybe overlaid) so we can understand what location in the warping field corresponds to the discontinuity in the segmentation.

Of course! I just updated [#11 (comment)](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/11#issuecomment-876534386)!

**Comment by jcohenadad** [comment]
ah! i think i know what this is. The Z component of the warping field (which is a 4D file) is "pushing" a voxel from outside (ie: defined as 0) inside the segmentation. A way to overcome this edge issue is simply by padding with mirroring the image when estimating the transformation.

**Comment by sandrinebedard** [comment]
Ahh I see! 

> Alternatively (easier) you can get rid of all edge slices.

Getting rid of the edge slices may be fine for T1w, but for T2star (same thing for MTS data and dwi), removing 2 slices is a bit extreme since we don't have a lot. 

I can try padding!

**Comment by sandrinebedard** [comment]
I tried padding with mirroring the image when estimating the transformation with `sct_register_multimodal`. Since the transformation is based on the segmentation, registration was worst for some images. 
e.g: 
sub-barcelona01_acq-T1w_MTS_reg.nii.gz with padding:
![anim](https://user-images.githubusercontent.com/71230552/125335641-c8a76300-e31a-11eb-912d-bb756af9de74.gif)
without padding:
![anim](https://user-images.githubusercontent.com/71230552/125335866-0c9a6800-e31b-11eb-9015-8405c7d3c773.gif)

I tried padding one slice over the segmentation and one under with the top and bottom slices. For now, it seams to correct the problem, but I will test with more images to make sure we don't have the problem anymore, in the case the edge effect affects more than one slice.

sub-barcelona01_acq-T1w_MTS_reg.nii.gz with padding of one slice:

![anim](https://user-images.githubusercontent.com/71230552/125336573-f17c2800-e31b-11eb-976b-8a7821ad1885.gif)

**Comment by jcohenadad** [comment]
How much padding did you do? I would start with only one slice at both extremities.

**Comment by sandrinebedard** [comment]
> How much padding did you do? I would start with only one slice at both extremities.

I tested out with 2 slicse at each at both extremities and then 1 : one slice at each at both extremities is what worked the best for regsitration and resolved the edge effect for the images I tested out!

The only thing I need to test out is for this case where 2 slices were affected if padding with 1 slice at each at both extremities resolves the problem!

![image](https://user-images.githubusercontent.com/71230552/125359243-ffd83d00-e337-11eb-9e8b-7b1ffd6178c2.png)

**Comment by sandrinebedard** [comment]
I tested out the image sub-tokyoSkyra_T1w that had 2 slices affected with the edge effect and with padding of 1 slice at each extremity, the problem is resolved 🎉

![image](https://user-images.githubusercontent.com/71230552/125479566-b3e33a3f-9edd-43a6-82a5-e39828951c00.png)

**Comment by joshuacwnewton** [comment]
> ah! i think i know what this is. The Z component of the warping field (which is a 4D file) is "pushing" a voxel from outside (ie: defined as 0) inside the segmentation. A way to overcome this edge effect is simply by padding with mirroring the image when estimating the transformation.

Have you encountered this in the past, @jcohenadad? 

I ask because I'm wondering if this behavior is common enough that we might want to check for and/or document it on SCT's end. (e.g. in `sct_apply_transfo`, checking the resulting transformed image for empty slices, then warning the user that padding might be helpful to fix the issue.)
